plugin-viewport/src/main/kotlin/com/mapbox/maps/plugin/viewport/experimental/ViewportPluginImpl.kt

```diff
@@ -0,0 +1,144 @@
+  }
+
+  override fun transitionToState(
+    state: ViewportState,
+    stateTransitionOptionsBlock: ViewportTransitionOptions.Builder.() -> Unit,
+    frameTransitionOptionsBlock: ViewportTransitionOptions.Builder.() -> Unit,
+    transitionEndListener: TransitionEndListener?
+  ) {
+    transitionToState(
+      state,
+      ViewportTransitionOptions.Builder().maxDuration(DEFAULT_STATE_TRANSITION_MAX_DURATION_MS)
+        .apply(stateTransitionOptionsBlock).build(),
+      ViewportTransitionOptions.Builder().maxDuration(DEFAULT_FRAME_TRANSITION_MAX_DURATION_MS)
+        .apply(frameTransitionOptionsBlock).build(),
+      transitionEndListener
+    )
+  }
+
+  override fun transitionToState(
+    state: ViewportState,
+    stateTransitionOptions: ViewportTransitionOptions,
+    frameTransitionOptions: ViewportTransitionOptions,
+    transitionEndListener: TransitionEndListener?
+  ) {
+    states.add(state)
+    currentState?.cleanUp()
+    state.transitionToState(
+      stateTransitionOptions,
+      frameTransitionOptions,
+      transitionEndListener
+    )
+    currentState = state
+    notifyStateChanged()
+  }
+
+  override fun transitionToState(
+    stateId: String,
+    stateTransitionOptionsBlock: ViewportTransitionOptions.Builder.() -> Unit,
+    frameTransitionOptionsBlock: ViewportTransitionOptions.Builder.() -> Unit,
+    transitionEndListener: TransitionEndListener?
+  ) {
+    transitionToState(
+      states.first { it.id == stateId },
+      stateTransitionOptionsBlock,
+      frameTransitionOptionsBlock,
+      transitionEndListener
+    )
+  }
+
+  override fun transitionToState(
+    stateId: String,
+    stateTransitionOptions: ViewportTransitionOptions,
+    frameTransitionOptions: ViewportTransitionOptions,
+    transitionEndListener: TransitionEndListener?
+  ) {
+    transitionToState(
+      states.first { it.id == stateId },
+      stateTransitionOptions,
+      frameTransitionOptions,
+      transitionEndListener
+    )
+  }
+
+  override fun transitionToIdle() {
+    currentState?.cleanUp()
+    currentState = null
+    notifyStateChanged()
+  }
+
+
+  private fun notifyStateChanged() {
+    registeredStateChangedObservers.forEach { it.onViewportStateChanged(currentState) }
+  }
+
+  override fun registerViewportStateChangedObserver(viewportStateChangedObserver: ViewportStateChangedObserver) {
+    registeredStateChangedObservers.add(viewportStateChangedObserver)
+  }
+
+  override fun unregisterViewportStateChangedObserver(viewportStateChangedObserver: ViewportStateChangedObserver) {
+    registeredStateChangedObservers.remove(viewportStateChangedObserver)
+  }
+
+  override val followingViewportState: FollowingViewportState
+    get() = states.first { it.id == ViewportState.MAPBOX_VIEWPORT_STATE_FOLLOWING } as FollowingViewportState
+
+  override val overviewViewportState: OverviewViewportState
+    get() = states.first { it.id == ViewportState.MAPBOX_VIEWPORT_STATE_OVERVIEW } as OverviewViewportState
```

Comment:
I don't think these properties should be part of the API. It's too specific to the built-in states. At a minimum, they need to be nullable since these states might not always exist.

Reply:
These are the two states provided by Mapbox by default, and we define these two interfaces to allow customising specific state settings.

Reply:
Do we need to register any states by default? Why not require that developers create and register instances of our default states?

Reply:
I think the idea is that we will provide Mapbox default implementation for some tracking modes.
Do you suggest we only provide state instance getters for the Mapbox built-in viewport states and ask them to add it to the plugin before they want to use it? Feels like some extra steps.

Reply:
And I think we'd like also to make the viewport plugin's mapbox presets(overview and following) very easy to use, since that would be the primary use case for most users.

Reply:
btw we need to initialise the FollowingViewportState and OverviewViewportState internally inside the viewport plugin, as they are dependent on some delegates from our plugin setup to hook into the camera animations/ location component.

Reply:
Updated the API according to the discussions offline, user would be able to create `FollowingViewportState` and `OverviewViewportState` instances and use them later on.